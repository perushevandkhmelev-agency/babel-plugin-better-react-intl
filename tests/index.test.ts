import {join, basename} from 'path';
import * as fs from 'fs';
import {transformFileSync} from '@babel/core';
import plugin, {resolveOutputPath} from '../';

function trim(str?: string | null) {
  return String(str).replace(/^\s+|\s+$/, '');
}

const skipOutputTests = [
  'additionalComponentNames',
  'empty',
  'extractFromFormatMessageCall',
  'extractFromFormatMessageCallStateless',
  'extractSourceLocation',
  'icuSyntax',
  'idInterpolationPattern',
  'inline',
  'moduleSourceName',
  'noMessagesDir',
  'outputEmptyJson',
  'overrideIdFn',
  'removeDefaultMessage',
  'removeDescriptions',
  'workspaceRoot',
];

describe('emit asserts for: ', () => {
  fs.readdirSync(join(__dirname, 'fixtures')).map(caseName => {
    if (skipOutputTests.indexOf(caseName) >= 0) return;

    it(`output match: ${caseName}`, () => {
      const fixtureDir = join(__dirname, 'fixtures', caseName);
      const filePath = join(fixtureDir, 'actual.js');

      const {code: actual, metadata} = transform(filePath, {
        pragma: '@react-intl',
      })!;
      expect((metadata as any)['react-intl']).toMatchSnapshot();
      // Check code output
      expect(trim(actual)).toMatchSnapshot();

      // Check message output
      expect(
        require(resolveOutputPath(process.cwd(), __dirname, filePath))
      ).toMatchSnapshot();
    });
  });
});

describe('options', () => {
  it('removeDefaultMessage should remove default message', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'removeDefaultMessage');
    const filePath = join(fixtureDir, 'actual.js');
    const actual = transform(filePath, {
      removeDefaultMessage: true,
    })!.code;

    // Check code output
    expect(trim(actual)).toMatchSnapshot();

    // Check message output
    expect(
      require(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toMatchSnapshot();
  });
  it('outputEmptyJson should output empty files', function () {
    const fixtureDir = join(__dirname, 'fixtures', 'outputEmptyJson');
    const filePath = join(fixtureDir, 'actual.js');
    const actual = transform(filePath, {
      outputEmptyJson: true,
    })!.code;

    // Check code output
    expect(trim(actual)).toMatchSnapshot();

    // Check message output
    expect(
      require(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toMatchSnapshot();
  });
  it('without outputEmptyJson should output empty files', function () {
    const fixtureDir = join(__dirname, 'fixtures', 'empty');
    const filePath = join(fixtureDir, 'actual.js');
    const actual = transform(filePath, {})!.code;

    // Check code output
    expect(trim(actual)).toMatchSnapshot();

    // Check message output
    expect(
      fs.existsSync(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toBeFalsy();
  });
  it('correctly overrides the id when overrideIdFn is provided', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'overrideIdFn');
    const filePath = join(fixtureDir, 'actual.js');
    const actual = transform(filePath, {
      overrideIdFn: (
        id: string,
        defaultMessage: string,
        description: string,
        filePath: string
      ) => {
        const filename = basename(filePath);
        return `${filename}.${id}.${
          defaultMessage.length
        }.${typeof description}`;
      },
    })!.code;

    // Check code output
    expect(trim(actual)).toMatchSnapshot();

    // Check message output
    expect(
      require(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toMatchSnapshot();
  });
  it('correctly overrides the id when idInterpolationPattern is provided', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'idInterpolationPattern');
    const filePath = join(fixtureDir, 'actual.js');
    const actual = transform(filePath, {
      idInterpolationPattern: '[sha512:contenthash:hex:6]',
    })!.code;

    // Check code output
    expect(trim(actual)).toMatchSnapshot();

    // Check message output
    expect(
      require(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toMatchSnapshot();
  });

  it('removes descriptions when plugin is applied more than once', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'removeDescriptions');
    expect(() =>
      transform(
        join(fixtureDir, 'actual.js'),
        {},
        {
          multiplePasses: true,
        }
      )
    ).not.toThrow();
  });

  it('respects moduleSourceName', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'moduleSourceName');
    const filePath = join(fixtureDir, 'actual.js');
    expect(() =>
      transform(filePath, {
        moduleSourceName: 'react-i18n',
      })
    ).not.toThrow();

    // Check message output
    expect(
      require(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toMatchSnapshot();
  });

  it('should be able to parse inline defineMessage from react-intl', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'inline');
    const filePath = join(fixtureDir, 'actual.js');
    expect(() => transform(filePath)).not.toThrow();

    // Check message output
    expect(
      require(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toMatchSnapshot();
  });

  it('respects extractSourceLocation', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'extractSourceLocation');
    const filePath = join(fixtureDir, 'actual.js');
    expect(() =>
      transform(filePath, {
        extractSourceLocation: true,
      })
    ).not.toThrow();

    // Check message output
    const actualMessages = require(resolveOutputPath(
      process.cwd(),
      __dirname,
      filePath
    ));
    actualMessages.forEach((msg: any) => {
      msg.file = msg.file.replace(/\/|\\/g, '@@sep@@');
    });
    expect(actualMessages).toMatchSnapshot();
  });

  it('respects extractFromFormatMessageCall', () => {
    const fixtureDir = join(
      __dirname,
      'fixtures',
      'extractFromFormatMessageCall'
    );
    const filePath = join(fixtureDir, 'actual.js');
    expect(() =>
      transform(filePath, {
        extractFromFormatMessageCall: true,
      })
    ).not.toThrow();

    // Check message output
    expect(
      require(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toMatchSnapshot();
  });

  it('respects extractFromFormatMessageCall from stateless components', () => {
    const fixtureDir = join(
      __dirname,
      'fixtures',
      'extractFromFormatMessageCallStateless'
    );
    const filePath = join(fixtureDir, 'actual.js');
    expect(() =>
      transform(filePath, {
        extractFromFormatMessageCall: true,
      })
    ).not.toThrow();

    // Check message output
    expect(
      require(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toMatchSnapshot();
  });

  it('additionalComponentNames', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'additionalComponentNames');
    const filePath = join(fixtureDir, 'actual.js');
    expect(() =>
      transform(filePath, {
        additionalComponentNames: ['CustomMessage'],
      })
    ).not.toThrow();

    // Check message output
    expect(
      require(resolveOutputPath(process.cwd(), __dirname, filePath))
    ).toMatchSnapshot();
  });

  it('undefined messagesDir should work normally (w/o writing file)', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'noMessagesDir');
    const filePath = join(fixtureDir, 'actual.js');
    const {code: actual, metadata} = transform(filePath, {
      pragma: '@react-intl',
      idInterpolationPattern: '[sha512:hash:base64:6]',
      messagesDir: undefined,
    })!;
    expect((metadata as any)['react-intl']).toMatchSnapshot();
    // Check code output
    expect(trim(actual)).toMatchSnapshot();
  });

  it('workspaceRoot', () => {
    const fixtureDir = join(
      __dirname,
      'fixtures',
      'workspaceRoot',
      'subdir1',
      'subdir2'
    );
    const filePath = join(fixtureDir, 'actual.js');
    expect(() =>
      transform(filePath, {
        workspaceRoot: join(__dirname, 'fixtures', 'workspaceRoot', 'subdir1'),
      })
    ).not.toThrow();

    // Check message output
    expect(
      require(resolveOutputPath(
        join(__dirname, 'fixtures', 'workspaceRoot', 'subdir1'),
        __dirname,
        filePath
      ))
    ).toMatchSnapshot();

    expect(() =>
      transform(filePath, {
        workspaceRoot: join(__dirname, 'fixtures', 'workspaceRoot'),
      })
    ).not.toThrow();

    // Check message output
    expect(
      require(resolveOutputPath(
        join(__dirname, 'fixtures', 'workspaceRoot'),
        __dirname,
        filePath
      ))
    ).toMatchSnapshot();
  });

  it('workspaceRoot invalid', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'workspaceRoot', 'subdir3');
    const filePath = join(fixtureDir, 'actual.js');
    expect(() =>
      transform(filePath, {
        workspaceRoot: join(__dirname, 'fixtures', 'workspaceRoot', 'subdir1'),
      })
    ).toThrow();
  });
});

describe('errors', () => {
  it('Properly throws parse errors', () => {
    const fixtureDir = join(__dirname, 'fixtures', 'icuSyntax');
    expect(() => transform(join(fixtureDir, 'actual.js'))).toThrow(
      /Expected .* but "\." found/
    );
  });
});

let cacheBust = 1;

function transform(
  filePath: string,
  options = {},
  {multiplePasses = false} = {}
) {
  function getPluginConfig() {
    return [
      plugin,
      {
        messagesDir: __dirname,
        ...options,
      },
      Date.now() + '' + ++cacheBust,
    ];
  }

  return transformFileSync(filePath, {
    plugins: multiplePasses
      ? [
          'module:@babel/plugin-syntax-jsx',
          getPluginConfig(),
          getPluginConfig(),
        ]
      : ['module:@babel/plugin-syntax-jsx', getPluginConfig()],
  });
}
