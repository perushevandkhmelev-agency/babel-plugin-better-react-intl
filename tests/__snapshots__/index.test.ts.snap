// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`emit asserts for:  output match: FormattedMessage 1`] = `
Object {
  "messages": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message.",
      "id": "foo.bar.baz",
    },
  ],
  "meta": Object {},
}
`;

exports[`emit asserts for:  output match: FormattedMessage 2`] = `
"import React, { Component } from 'react';
import { FormattedMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage=\\"Hello World!\\" />;
  }

}"
`;

exports[`emit asserts for:  output match: FormattedMessage 3`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message.",
    "id": "foo.bar.baz",
  },
]
`;

exports[`emit asserts for:  output match: defineMessage 1`] = `
Object {
  "messages": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "foo.bar.baz",
    },
    Object {
      "defaultMessage": "Hello Nurse!",
      "description": "Another message",
      "id": "foo.bar.biff",
    },
    Object {
      "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
      "description": "Counts kittens",
      "id": "app.home.kittens",
    },
    Object {
      "defaultMessage": "   Some whitespace   ",
      "description": "Whitespace",
      "id": "trailing.ws",
    },
    Object {
      "defaultMessage": "A quoted value ''{value}'",
      "description": "Escaped apostrophe",
      "id": "escaped.apostrophe",
    },
  ],
  "meta": Object {
    "project": "amazing",
  },
}
`;

exports[`emit asserts for:  output match: defineMessage 2`] = `
"// @react-intl project:amazing
import React, { Component } from 'react';
import { defineMessage, FormattedMessage } from 'react-intl';
const msgs = {
  header: defineMessage({
    \\"id\\": \\"foo.bar.baz\\",
    \\"defaultMessage\\": \\"Hello World!\\"
  }),
  content: defineMessage({
    \\"id\\": \\"foo.bar.biff\\",
    \\"defaultMessage\\": \\"Hello Nurse!\\"
  }),
  kittens: defineMessage({
    \\"id\\": \\"app.home.kittens\\",
    \\"defaultMessage\\": \\"{count, plural, =0 {\\\\uD83D\\\\uDE2D} one {# kitten} other {# kittens}}\\"
  }),
  trailingWhitespace: defineMessage({
    \\"id\\": \\"trailing.ws\\",
    \\"defaultMessage\\": \\"   Some whitespace   \\"
  }),
  escaped: defineMessage({
    \\"id\\": \\"escaped.apostrophe\\",
    \\"defaultMessage\\": \\"A quoted value ''{value}'\\"
  })
};
export default class Foo extends Component {
  render() {
    return <div>
        <h1>
          <FormattedMessage {...msgs.header} />
        </h1>
        <p>
          <FormattedMessage {...msgs.content} />
        </p>
        <p>
          <FormattedMessage {...msgs.kittens} />
        </p>
      </div>;
  }

}"
`;

exports[`emit asserts for:  output match: defineMessage 3`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
    "id": "app.home.kittens",
  },
  Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
    "id": "trailing.ws",
  },
  Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
    "id": "escaped.apostrophe",
  },
]
`;

exports[`emit asserts for:  output match: defineMessages 1`] = `
Object {
  "messages": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "The default message",
      "id": "foo.bar.baz",
    },
    Object {
      "defaultMessage": "Hello Nurse!",
      "description": "Another message",
      "id": "foo.bar.biff",
    },
    Object {
      "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
      "description": "Counts kittens",
      "id": "app.home.kittens",
    },
    Object {
      "defaultMessage": "   Some whitespace   ",
      "description": "Whitespace",
      "id": "trailing.ws",
    },
    Object {
      "defaultMessage": "A quoted value ''{value}'",
      "description": "Escaped apostrophe",
      "id": "escaped.apostrophe",
    },
  ],
  "meta": Object {
    "project": "amazing",
  },
}
`;

exports[`emit asserts for:  output match: defineMessages 2`] = `
"// @react-intl project:amazing
import React, { Component } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
const msgs = defineMessages({
  header: {
    \\"id\\": \\"foo.bar.baz\\",
    \\"defaultMessage\\": \\"Hello World!\\"
  },
  content: {
    \\"id\\": \\"foo.bar.biff\\",
    \\"defaultMessage\\": \\"Hello Nurse!\\"
  },
  kittens: {
    \\"id\\": \\"app.home.kittens\\",
    \\"defaultMessage\\": \\"{count, plural, =0 {\\\\uD83D\\\\uDE2D} one {# kitten} other {# kittens}}\\"
  },
  trailingWhitespace: {
    \\"id\\": \\"trailing.ws\\",
    \\"defaultMessage\\": \\"   Some whitespace   \\"
  },
  escaped: {
    \\"id\\": \\"escaped.apostrophe\\",
    \\"defaultMessage\\": \\"A quoted value ''{value}'\\"
  }
});
export default class Foo extends Component {
  render() {
    return <div>
        <h1>
          <FormattedMessage {...msgs.header} />
        </h1>
        <p>
          <FormattedMessage {...msgs.content} />
        </p>
        <p>
          <FormattedMessage {...msgs.kittens} />
        </p>
      </div>;
  }

}"
`;

exports[`emit asserts for:  output match: defineMessages 3`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "{count, plural, =0 {ðŸ˜­} one {# kitten} other {# kittens}}",
    "description": "Counts kittens",
    "id": "app.home.kittens",
  },
  Object {
    "defaultMessage": "   Some whitespace   ",
    "description": "Whitespace",
    "id": "trailing.ws",
  },
  Object {
    "defaultMessage": "A quoted value ''{value}'",
    "description": "Escaped apostrophe",
    "id": "escaped.apostrophe",
  },
]
`;

exports[`emit asserts for:  output match: descriptionsAsObjects 1`] = `
Object {
  "messages": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": Object {
        "metadata": "Additional metadata content.",
        "text": "Something for the translator.",
      },
      "id": "foo.bar.baz",
    },
  ],
  "meta": Object {
    "project": "amazing2",
  },
}
`;

exports[`emit asserts for:  output match: descriptionsAsObjects 2`] = `
"import React, { Component } from 'react';
import { FormattedMessage } from 'react-intl'; // @react-intl project:amazing2

export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage=\\"Hello World!\\" />;
  }

}"
`;

exports[`emit asserts for:  output match: descriptionsAsObjects 3`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": Object {
      "metadata": "Additional metadata content.",
      "text": "Something for the translator.",
    },
    "id": "foo.bar.baz",
  },
]
`;

exports[`emit asserts for:  output match: enforceDefaultMessage 1`] = `
Object {
  "messages": Array [
    Object {
      "defaultMessage": undefined,
      "description": undefined,
      "id": "TRANSLATION_KEY",
    },
    Object {
      "defaultMessage": undefined,
      "description": undefined,
      "id": "foo.bar.baz",
    },
  ],
  "meta": Object {},
}
`;

exports[`emit asserts for:  output match: enforceDefaultMessage 2`] = `
"import React, { Component } from 'react';
import { FormattedMessage, defineMessages } from 'react-intl';
defineMessages({
  test: {
    \\"id\\": \\"TRANSLATION_KEY\\"
  }
});
const testMessage = {
  id: 'TRANSLATION_KEY2'
};
export default class Foo extends Component {
  render() {
    return <>
        <FormattedMessage id=\\"foo.bar.baz\\" />
        <FormattedMessage {...testMessage} />
      </>;
  }

}"
`;

exports[`emit asserts for:  output match: enforceDefaultMessage 3`] = `
Array [
  Object {
    "id": "TRANSLATION_KEY",
  },
  Object {
    "id": "foo.bar.baz",
  },
]
`;

exports[`emit asserts for:  output match: enforceDescriptions 1`] = `
Object {
  "messages": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": undefined,
      "id": "foo.bar.baz",
    },
  ],
  "meta": Object {},
}
`;

exports[`emit asserts for:  output match: enforceDescriptions 2`] = `
"import React, { Component } from 'react';
import { FormattedMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"foo.bar.baz\\" defaultMessage=\\"Hello World!\\" />;
  }

}"
`;

exports[`emit asserts for:  output match: enforceDescriptions 3`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "id": "foo.bar.baz",
  },
]
`;

exports[`emit asserts for:  output match: formatMessageCall 1`] = `
Object {
  "messages": Array [
    Object {
      "defaultMessage": "bar",
      "description": "baz",
      "id": "foo",
    },
  ],
  "meta": Object {},
}
`;

exports[`emit asserts for:  output match: formatMessageCall 2`] = `
"import React, { Component } from 'react';
import { injectIntl, FormattedMessage } from 'react-intl';
const objectPointer = {
  id: 'foo.bar.invalid',
  defaultMessage: 'This cannot be extracted',
  description: 'the plugin only supports inline objects'
};

class Foo extends Component {
  render() {
    const msgs = {
      baz: this.props.intl.formatMessage({
        id: 'foo.bar.baz',
        defaultMessage: 'Hello World!',
        description: 'The default message'
      }),
      biff: this.props.intl.formatMessage({
        id: 'foo.bar.biff',
        defaultMessage: 'Hello Nurse!',
        description: 'Another message'
      }),
      invalid: this.props.intl.formatMessage(objectPointer)
    };
    return <div>
        <h1>{msgs.header}</h1>
        <p>{msgs.content}</p>
        <span>
          <FormattedMessage id=\\"foo\\" defaultMessage=\\"bar\\" />
        </span>
      </div>;
  }

}

export default injectIntl(Foo);"
`;

exports[`emit asserts for:  output match: formatMessageCall 3`] = `
Array [
  Object {
    "defaultMessage": "bar",
    "description": "baz",
    "id": "foo",
  },
]
`;

exports[`options additionalComponentNames 1`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "Greeting to the world",
    "id": "greeting-world",
  },
]
`;

exports[`options correctly overrides the id when idInterpolationPattern is provided 1`] = `
"import React, { Component } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
const msgs = defineMessages({
  header: {
    \\"id\\": \\"4d1460\\",
    \\"defaultMessage\\": \\"Hello World!\\"
  },
  content: {
    \\"id\\": \\"foo.bar.biff\\",
    \\"defaultMessage\\": \\"Hello Nurse!\\"
  }
});
export default class Foo extends Component {
  render() {
    return <div>
        <h1>
          <FormattedMessage {...msgs.header} />
        </h1>
        <p>
          <FormattedMessage {...msgs.content} />
        </p>
        <FormattedMessage id=\\"36a8c8\\" defaultMessage=\\"Hello World!\\" />
        <FormattedMessage id=\\"5ce864\\" defaultMessage=\\"NO ID\\" />
      </div>;
  }

}"
`;

exports[`options correctly overrides the id when idInterpolationPattern is provided 2`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "4d1460",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": Object {
      "metadata": "Additional metadata content.",
      "text": "Something for the translator.",
    },
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "Hello World!",
    "description": "Something for the translator. Another description",
    "id": "36a8c8",
  },
  Object {
    "defaultMessage": "NO ID",
    "description": "Something for the translator. Another description",
    "id": "5ce864",
  },
]
`;

exports[`options correctly overrides the id when overrideIdFn is provided 1`] = `
"import React, { Component } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
const msgs = defineMessages({
  header: {
    \\"id\\": \\"actual.js.foo.bar.baz.12.string\\",
    \\"defaultMessage\\": \\"Hello World!\\"
  },
  content: {
    \\"id\\": \\"actual.js.foo.bar.biff.12.object\\",
    \\"defaultMessage\\": \\"Hello Nurse!\\"
  }
});
export default class Foo extends Component {
  render() {
    return <div>
        <h1>
          <FormattedMessage {...msgs.header} />
        </h1>
        <p>
          <FormattedMessage {...msgs.content} />
        </p>
        <FormattedMessage id=\\"actual.js.foo.bar.zoo.12.object\\" defaultMessage=\\"Hello World!\\" />
        <FormattedMessage id=\\"actual.js..5.object\\" defaultMessage=\\"NO ID\\" />
      </div>;
  }

}"
`;

exports[`options correctly overrides the id when overrideIdFn is provided 2`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "actual.js.foo.bar.baz.12.string",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": Object {
      "metadata": "Additional metadata content.",
      "text": "Something for the translator.",
    },
    "id": "actual.js.foo.bar.biff.12.object",
  },
  Object {
    "defaultMessage": "Hello World!",
    "description": Object {
      "metadata": "Additional metadata content.",
      "text": "Something for the translator. Another description",
    },
    "id": "actual.js.foo.bar.zoo.12.object",
  },
  Object {
    "defaultMessage": "NO ID",
    "description": Object {
      "metadata": "Additional metadata content.",
      "text": "Something for the translator. Another description",
    },
    "id": "actual.js..5.object",
  },
]
`;

exports[`options outputEmptyJson should output empty files 1`] = `
"import React, { Component } from 'react';
import { defineMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <div></div>;
  }

}"
`;

exports[`options outputEmptyJson should output empty files 2`] = `Array []`;

exports[`options removeDefaultMessage should remove default message 1`] = `
"import React, { Component } from 'react';
import { defineMessages, FormattedMessage } from 'react-intl';
const messages = defineMessages({
  foo: {
    \\"id\\": \\"greeting-user\\"
  }
});
export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"greeting-world\\" />;
  }

}"
`;

exports[`options removeDefaultMessage should remove default message 2`] = `
Array [
  Object {
    "defaultMessage": "Hello, {name}",
    "description": "Greeting the user",
    "id": "greeting-user",
  },
  Object {
    "defaultMessage": "Hello World!",
    "description": "Greeting to the world",
    "id": "greeting-world",
  },
]
`;

exports[`options respects extractFromFormatMessageCall 1`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello Nurse!",
    "description": "Another message",
    "id": "foo.bar.biff",
  },
  Object {
    "defaultMessage": "Hello Stranger!",
    "description": "A different message",
    "id": "foo.bar.qux",
  },
  Object {
    "defaultMessage": "bar",
    "description": "baz",
    "id": "foo",
  },
]
`;

exports[`options respects extractFromFormatMessageCall from stateless components 1`] = `
Array [
  Object {
    "defaultMessage": "hook <b>foo</b>",
    "description": "hook",
    "id": "hook",
  },
  Object {
    "defaultMessage": "bar",
    "description": "baz",
    "id": "foo",
  },
]
`;

exports[`options respects extractSourceLocation 1`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "end": Object {
      "column": 78,
      "line": 6,
    },
    "file": "packages@@sep@@babel-plugin-react-intl@@sep@@tests@@sep@@fixtures@@sep@@extractSourceLocation@@sep@@actual.js",
    "id": "foo.bar.baz",
    "start": Object {
      "column": 11,
      "line": 6,
    },
  },
]
`;

exports[`options respects moduleSourceName 1`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message.",
    "id": "foo.bar.baz",
  },
]
`;

exports[`options should be able to parse inline defineMessage from react-intl 1`] = `
Array [
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message.",
    "id": "foo.bar.baz",
  },
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "header",
  },
  Object {
    "defaultMessage": "Hello World!",
    "description": "The default message",
    "id": "header2",
  },
]
`;

exports[`options undefined messagesDir should work normally (w/o writing file) 1`] = `
Object {
  "messages": Array [
    Object {
      "defaultMessage": "Hello World!",
      "description": "Greeting to the world",
      "id": "cvAgVK",
    },
  ],
  "meta": Object {},
}
`;

exports[`options undefined messagesDir should work normally (w/o writing file) 2`] = `
"import React, { Component } from 'react';
import { FormattedMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <FormattedMessage id=\\"cvAgVK\\" defaultMessage=\\"Hello World!\\" />;
  }

}"
`;

exports[`options without outputEmptyJson should output empty files 1`] = `
"import React, { Component } from 'react';
import { defineMessage } from 'react-intl';
export default class Foo extends Component {
  render() {
    return <div></div>;
  }

}"
`;

exports[`options workspaceRoot 1`] = `
Array [
  Object {
    "defaultMessage": "Hello, {name}",
    "description": "Greeting the user",
    "id": "greeting-user",
  },
  Object {
    "defaultMessage": "Hello World!",
    "description": "Greeting to the world",
    "id": "greeting-world",
  },
]
`;

exports[`options workspaceRoot 2`] = `
Array [
  Object {
    "defaultMessage": "Hello, {name}",
    "description": "Greeting the user",
    "id": "greeting-user",
  },
  Object {
    "defaultMessage": "Hello World!",
    "description": "Greeting to the world",
    "id": "greeting-world",
  },
]
`;
